src/assets/TokenListController.ts

```diff
@@ -126,29 +135,38 @@ export class TokenListController extends BaseController<
     this.chainId = chainId;
     this.updatePreventPollingOnNetworkRestart(preventPollingOnNetworkRestart);
     this.abortController = new AbortController();
-    this.messagingSystem.subscribe(
-      'NetworkController:providerChange',
-      async (providerConfig) => {
-        if (this.chainId !== providerConfig.chainId) {
-          this.abortController.abort();
-          this.abortController = new AbortController();
-          this.chainId = providerConfig.chainId;
-          if (this.state.preventPollingOnNetworkRestart) {
-            this.clearingTokenListData();
-          } else {
-            // Ensure tokenList is referencing data from correct network
-            this.update(() => {
-              return {
-                ...this.state,
-                tokenList:
-                  this.state.tokensChainsCache[this.chainId]?.data || {},
-              };
-            });
-            await this.restart();
-          }
-        }
-      },
-    );
+    if (onNetworkStateChange) {
+      onNetworkStateChange(async ({ provider }) => {
+        await this.onNetworkStateChangeCb(provider);
+      });
+    } else {
+      this.messagingSystem.subscribe(
+        'NetworkController:providerChange',
+        async (providerConfig) => {
+          await this.onNetworkStateChangeCb(providerConfig);
+        },
+      );
+    }
+  }
+
+  async onNetworkStateChangeCb(providerConfig: ProviderConfig) {
```

Comment:
Nit: could we expand `Cb` to `Callback`? Would like to avoid abbreviations if possible, they can be harder to read

Reply:
Done [here](https://github.com/MetaMask/controllers/pull/932/commits/29020fdd1d80e00cab032adcadc71efaa9e1b1ec)